feat(cancel): add cancel-in-progress feature

chmouel opened the pull request:
# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

Implemented the cancel-in-progress feature to allow cancellation of
ongoing PipelineRuns associated with a repository and pull request.
This feature checks for concurrency limits and handles conflicts
robustly. It improves the management of PipelineRuns by ensuring
unnecessary runs can be cancelled efficiently.

Jira: <https://issues.redhat.com/browse/SRVKP-6745>
Demo: https://youtu.be/ayqtVG8cEnY
Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

# Submitter Checklist

- [x] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).
   x
- [x] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.
   x
- [x] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).
   x
- [x] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.
   x
- [x] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.
   x
- [x] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.
   x
- [x] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

chmouel commented:
/cancel go-testing

chmouel commented:
/test go-testing

vdemeester commented on pkg/pipelineascode/cancel_pipelineruns.go:
```diff
@@ -24,6 +24,63 @@ var cancelMergePatch = map[string]interface{}{
+	if !ok {
+		return nil
+	}
+	labelSelector := getLabelSelector(map[string]string{
+		keys.URLRepository:  formatting.CleanValueKubernetes(p.event.Repository),
+		keys.OriginalPRName: prName,
```
`OriginalPRName` is the name of the PipelineRun in the repository right ? (prior to have a "generated" part for it)

chmouel replied:
yah

vdemeester reviewed (approved):
❤️

chmouel commented:
Good catch by E2E, i am getting this errors:

<pre>
2024-11-15T09:11:57.810512034Z stdout F {"level":"error","ts":"2024-11-15T09:11:57.810Z","logger":"pipelinesascode","caller":"events/emit.go:46","msg":"There was an error starting the PipelineRun pr-bad-format-zqqz-pac-e2e-test-pzj2k-, creating pipelinerun pr-bad-format-zqqz-pac-e2e-test-pzj2k- in namespace pac-e2e-test-pzj2k has failed.\n\nTekton Controller has reported this error: ```admission webhook \"validation.webhook.pipeline.tekton.dev\" denied the request: validation failed: expected exactly one, got neither: spec.pipelineRef, spec.pipelineSpec``` ","commit":"b61e5aa","provider":"gitea","event-id":"c55fd491-c11e-4310-9f56-d0bee0c75f3e","event-sha":"4584fbdc62a813192df63538aff03ec23bf11c6f","event-type":"pull_request","namespace":"pac-e2e-test-pzj2k","stacktrace":"github.com/openshift-pipelines/pipelines-as-code/pkg/events.(*EventEmitter).EmitMessage\n\tgithub.com/openshift-pipelines/pipelines-as-code/pkg/events/emit.go:46\ngithub.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode.(*PacRun).Run.func1\n\tgithub.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode/pipelineascode.go:106"}
2024-11-15T09:11:57.856386335Z stderr F panic: runtime error: invalid memory address or nil pointer dereference
2024-11-15T09:11:57.856401424Z stderr F [signal SIGSEGV: segmentation violation code=0x1 addr=0xb8 pc=0x20cb805]
2024-11-15T09:11:57.856407215Z stderr F 
2024-11-15T09:11:57.856411503Z stderr F goroutine 2636 [running]:
2024-11-15T09:11:57.856417063Z stderr F github.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode.(*PacRun).cancelInProgress(0xc00074ab40, {0x2cb9518, 0xc00023d800}, 0x0, 0xc000b96c60)
2024-11-15T09:11:57.856437382Z stderr F 	github.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode/cancel_pipelineruns.go:34 +0x45
2024-11-15T09:11:57.85645796Z stderr F github.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode.(*PacRun).Run.func1({0xc00045d448?, 0xc0010eda20?, 0xc000b58900?}, 0x0)
2024-11-15T09:11:57.856464122Z stderr F 	github.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode/pipelineascode.go:119 +0x4d2
2024-11-15T09:11:57.856469863Z stderr F created by github.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode.(*PacRun).Run in goroutine 2676
2024-11-15T09:11:57.856475253Z stderr F 	github.com/openshift-pipelines/pipelines-as-code/pkg/pipelineascode/pipelineascode.go:100 +0x56a
</details>
</pre>

chmouel commented:
/retest

chmouel committed:
feat: allow cancelling in-progress PipelineRuns

- Added new annotation `pipelinesascode.tekton.dev/cancel-in-progress: "true"`
  to allow cancelling in-progress PipelineRuns.
- Updated documentation to include instructions for cancelling in-progress
  PipelineRuns.
- Implemented `cancelInProgress` function to handle the cancellation logic.
- Added unit tests for the new feature.
- Modified existing functions to support the new annotation.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>
